specify/specify7
# Add better support for one-to-one relationships (COG/COJO Data Entry fixes)

Fixes #5508
Fixes #5452
Fixes #5461
Somewhat Fixes #5463 (Save blocker is handled like any other required field: removing the COJO lets the user save) 

Mostly a reimplementation of https://github.com/specify/specify7/pull/5509 without serious consideration of Clone/CarryForward behavior

### Checklist

      self-explanatory (or properly documented)

### Testing instructions
For #5508: 
- Create or find a Collection Object Group 
- Add a child (COJO) to the Collection Object Group (COG) and save the COG
  - This can be either type of COJO child: Collection Object (CO) or another Collection Object Group
- [ ] Expand the child form on the COG (either CO or other COG), and ensure that changes and save-blockers "bubble" up to the primary COG form
  - i.e., make a change in a child form and ensure that you can save the parent form, and that any save blockers (like a uniqueness rule violation) prevents you from saving the parent form
- [ ] Save the COG and make sure that changes to any children are still bubbling up to COG form
- If not on the forms already, add the `cojo` relationship as a Subview to the CO or COG "base" Data Entry forms. 
- [ ] Go to the Data Entry form of a COJO child on the COG (either CO or COG) and make sure that the COJO displays correctly and changes to the COJO (and all related records to the COJO, i.e., COJO -> parentCog) propagate to the CO or COG. 
- [ ]  Save the CO or COG and make sure that changes to the COJO still propagate to the CO or COG

For #5452: 
- Create or find a Collection Object Group 
- Add a child (COJO) to the Collection Object Group (COG) and save the COG
  - This can be either type of COJO child: Collection Object (CO) or another Collection Object Group
- [ ] Go to the main Data Entry form of the COJO child you added to the COG (CO or COG), make a change, and save the form
- [ ] Go back the COG you created/found in step one (refresh the page if needed),  and make sure child is still associated with the COG

Please also do some general testing of subviews, both Independent and Dependent 🙏 : 
- Especially make sure the first Issue from https://github.com/specify/specify7/pull/5253#issuecomment-2334425790 can not be recreated
  - when editing both sides of a relationship as a subview (e.g., `CO -> accession` and `Accession -> collectionObjects`)
  - when making an edit when "distant" from the base form (e.g., making an edit in the last relationship of `CollectionObject -> determinations -> determiners -> agent`)
- It would also be nice to test the Issues which were mentioned as fixed in https://github.com/specify/specify7/pull/5253#issuecomment-2344003292!

Clone + Carry-Forward (from https://github.com/specify/specify7/pull/5509#issue-2753214588): 

Test Cloning:
- [ ] Select a CO form and add a parentCog to it in the Cojo subform.
- [ ] Save the CO
- [ ] Press the Clone button (no errors occur)
- [ ] Verify that you can  save the new cloned CO and that the parent is present 
- [ ] Verify it works for carry forward 
- [ ] Verify it works for bulk carry
